E2EReporter.ts

```diff
@@ -0,0 +1,207 @@
+      title: test.parent.title,
+      fileName: test.parent.parent!.title,
+      duration: 0,
+      tests: [],
+      suiteId: suiteId,
+    };
+
+    if (!suiteResult) {
+      this.suites.push(data);
+    }
+  }
+
+  onTestEnd(test: TestCase, result: TestResult) {
+    // find the suite this test belongs to
+    const suiteId = test.parent.title.split(" ").join(" ");
+    const suiteResult = this.suites.find((s) => s.suiteId === suiteId);
+
+    // @ts-ignore - using _projectId as projectId is not in the testCase types
+    const browserName = test._projectId as string;
+
+    // build test data
+    const data: CustomTestResult = {
+      title: test.title,
+      status: result.status,
+      duration: result.duration,
+      retries: result.retry,
+      errors: result.errors,
+      browser: browserName,
+    };
+
+    // add this to test to its suite
+    suiteResult?.tests.push(data);
+
+    // build test stats
+    switch (result.status) {
+      case "passed":
+        this.tempData.testsPassed++;
+        break;
+
+      case "failed":
+        this.tempData.testsFailed++;
+        this.tempData.failures.push(data);
+        break;
+
+      case "timedOut":
+        this.tempData.testsTimedout++;
+        this.tempData.failures.push(data);
+        break;
+
+      case "interrupted":
+        this.tempData.testsTimedout++;
+        this.tempData.failures.push(data);
+        break;
+
+      case "skipped":
+        this.tempData.testsSkipped++;
+        this.tempData.skipped.push(data);
+        break;
+
+      default:
+        this.tempData.testsFlaky++;
+        this.tempData.flakes.push(data);
+        break;
+    }
+
+    // increase test count
+    this.tempData.totalTests++;
+
+    // add duration
+    this.tempData.duration = this.tempData.duration + result.duration;
+  }
+
+  onEnd(result: FullResult) {
+    this.endTime = Date.now();
+
+    // build report
+    const report: CustomReport = {
+      id: Date.now(),
+      suites: this.suites,
+      totalDuration: result.duration,
+      startTime: new Date(this.startTime!),
+      endTime: new Date(this.endTime),
+      totalTests: this.tempData.totalTests,
+      testsPassed: this.tempData.testsPassed,
+      testsFailed: this.tempData.testsFailed,
+      testsFlaky: this.tempData.testsFlaky,
+      testsSkipped: this.tempData.testsSkipped,
+      averageTestDuration: result.duration / this.tempData.totalTests,
+      testsTimedout: this.tempData.testsTimedout,
+      status: result.status,
+      environment: process.env.E2E_TEST_ENV as string,
+      failures: this.tempData.failures,
+      skipped: this.tempData.skipped,
+      flakes: this.tempData.flakes,
+    };
+
+    // prepare the different json report
+    const reportJson = JSON.stringify(report, null, 2);
+
+    // flaky tests
+    const flakeJson = JSON.stringify(this.tempData.flakes, null, 2);
+
+    // failed tests
+    const failedJson = JSON.stringify(this.tempData.failures, null, 2);
```

Comment:
not sure if we want the `, null, 2` pretty-print business or if we just use `JSON.stringify(abc)` and write out minified logs and save 20-30% disk space.